google-cloud-storage/src/main/scala/akka/stream/alpakka/googlecloud/storage/impl/GoogleCloudStorageClient.scala

```diff
@@ -0,0 +1,459 @@
+      createAuthenticatedRequest(
+        baseUri
+          .withPath(basePath ++ sourcePath / "rewriteTo" ++ destinationPath)
+          .withQuery(
+            Query(rewriteToken.map(token => Map("rewriteToken" -> token)).getOrElse(Map()))
+          ),
+        HttpMethods.POST
+      ).flatMap(Http().singleRequest(_))
+        .flatMap(entityForSuccess)
+        .flatMap(responseEntityTo[RewriteResponse])
+        .map { rewriteResponse =>
+          Some(
+            rewriteResponse.rewriteToken.fold[(RewriteState, RewriteResponse)]((Finished, rewriteResponse))(
+              token => (Running(token), rewriteResponse)
+            )
+          )
+        }
+
+    Source
+      .unfoldAsync[RewriteState, RewriteResponse](Starting) {
+        case Finished => Future.successful(None)
+        case Starting => rewriteRequest(None)
+        case Running(token) => rewriteRequest(Some(token))
+      }
+      .runWith(Sink.last[RewriteResponse])
+      .map(_.resource)
+      .map(_.getOrElse(throw new RuntimeException("Storage object is missing")))
+  }
+
+  private def postRequest(
+      fullPath: Path,
+      queryParams: Map[String, String],
+      contentType: ContentType,
+      bytes: ByteString
+  ): Future[HttpResponse] =
+    createAuthenticatedRequest(baseUri.withPath(fullPath).withQuery(Query(queryParams)), HttpMethods.POST)
+      .flatMap(req => Http().singleRequest(req.withEntity(contentType, bytes)))
+
+  private def postSourceRequest(
+      fullPath: Path,
+      queryParams: Map[String, String],
+      contentType: ContentType,
+      data: Source[ByteString, _]
+  ): Future[HttpResponse] =
+    createAuthenticatedRequest(baseUri.withPath(fullPath).withQuery(Query(queryParams)), HttpMethods.POST)
+      .flatMap(req => Http().singleRequest(req.withEntity(HttpEntity(contentType, data))))
+
+  private def postRequestWithHeaders(
+      fullPath: Path,
+      queryParams: Map[String, String],
+      headers: Seq[HttpHeader]
+  ): Future[HttpResponse] =
+    createAuthenticatedRequest(baseUri.withPath(fullPath).withQuery(Query(queryParams)), HttpMethods.POST, headers)
+      .flatMap(Http().singleRequest(_))
+
+  private def request(
+      path: Path,
+      method: HttpMethod = HttpMethods.GET,
+      queryParams: Map[String, String] = Map.empty
+  ): Future[HttpResponse] =
+    createAuthenticatedRequest(baseUri.withPath(basePath ++ path).withQuery(Query(queryParams)), method)
+      .flatMap(Http().singleRequest(_))
+
+  private def createAuthenticatedRequest(uri: Uri,
+                                         method: HttpMethod,
+                                         headers: Seq[HttpHeader] = Seq.empty): Future[HttpRequest] =
+    session.getToken().map { accessToken =>
+      HttpRequest(method)
+        .withUri(uri)
+        .withHeaders(headers: _*)
+        .addCredentials(OAuth2BearerToken(accessToken))
+    }
+
+  private def getBucketPath(bucket: String) =
+    Path("/b") / bucket
```

Comment:
Yes it does, I'll add validations for bucket and object names. Should I throw exception on invalid value or return some kind of error?

Reply:
I went through this documentation https://cloud.google.com/storage/docs/naming and to me it seems that we don't need to implement validations since we will return google error messages to user as part of exception message

Here https://cloud.google.com/storage/docs/json_api/#encoding it says that object names should be encoded so I'll add it.